core/services/pipeline/orm.go

```diff
@@ -209,22 +248,16 @@ func (o *orm) processNextUnclaimedTaskRun(ctx context.Context, fn ProcessTaskRun
 
 		// Find (and lock) the next unlocked, unfinished pipeline_task_run with no uncompleted predecessors
 		err := tx.Raw(`
-            SELECT * from pipeline_task_runs WHERE id IN (
-                SELECT pipeline_task_runs.id FROM pipeline_task_runs
-                    INNER JOIN pipeline_task_specs ON pipeline_task_runs.pipeline_task_spec_id = pipeline_task_specs.id
-                    LEFT JOIN pipeline_task_specs AS predecessor_specs ON predecessor_specs.successor_id = pipeline_task_specs.id
-                    LEFT JOIN pipeline_task_runs AS predecessor_unfinished_runs ON predecessor_specs.id = predecessor_unfinished_runs.pipeline_task_spec_id
-                          AND pipeline_task_runs.pipeline_run_id = predecessor_unfinished_runs.pipeline_run_id
-                WHERE pipeline_task_runs.finished_at IS NULL
-                GROUP BY (pipeline_task_runs.id)
-                HAVING (
-                    bool_and(predecessor_unfinished_runs.finished_at IS NOT NULL)
-                    OR
-                    count(predecessor_unfinished_runs.id) = 0
-                )
-            )
-            LIMIT 1
-            FOR UPDATE OF pipeline_task_runs SKIP LOCKED;
+			SELECT * FROM pipeline_task_runs WHERE id IN (
+				SELECT runs.id FROM pipeline_task_runs AS runs
+				LEFT JOIN pipeline_task_runs AS preds ON preds.id = ANY (runs.predecessor_task_run_ids)
+				WHERE runs.finished_at is null
+				GROUP BY runs.id
+				HAVING (bool_and(preds.finished_at is not null) 
+					OR count(preds) = 0)
+			)
+           LIMIT 1
+           FOR UPDATE OF pipeline_task_runs SKIP LOCKED;
```

Comment:
This is much better. I'm kicking myself for not thinking of it before, I must have spent so much time staring at this query I forgot the obvious.